### PR TITLE
167 refactor encryption

### LINKED ISSUE #167: Refactor Encryption to Streams
Decorator pattern ftw!!

### PR DESCRIPTION
**Pull request by MathiasMeuleman**
Will close #167 

This PR features the refactoring of the JSON security.
The encrypt and decrypt methods in Parser and Writer are gone.
Instead, Parser and Writer now use a Base64Reader and Base64Writer, respecively.
These classes take care of the encryption and decryption of JSON internally before passing the read/write operation through to a BufferedReader/Writer.

Enjoy reviewing :)

<review

### CODE REVIEW

File: src/main/java/nl/tudelft/ti2206/group9/level/save/Parser.java [hunk 1]
```diff
@@ -59,20 +54,14 @@ private Parser() { }
...
-            final BufferedReader reader = new BufferedReader(
-                    new InputStreamReader(stream, "UTF-8"));
-
-            while (reader.ready()) {
-                lines.add(reader.readLine());
-            }
-            final String mainString = createString(lines);
-            final String decryptedMain = decrypt(mainString);
+            final Base64Reader reader = new Base64Reader(new BufferedReader(
+                    new InputStreamReader(stream, "UTF-8")));
```
Comment (mpsijm) [inline src/main/java/nl/tudelft/ti2206/group9/level/save/Parser.java]: Wouldn't it be easier to create a factory method in `Base64Reader` (maybe also in Writer) that creates a `Base64Reader` from an InputStream? That would greatly enhance the readability of these lines.

<review, logic, interface

File: src/main/java/nl/tudelft/ti2206/group9/util/Base64Reader.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+public class Base64Reader extends Reader {
+
+    /** The Reader you decorate with this Base64Reader. */
+    private final Reader reader;
+    /** The Base64 decoder. */
+    private final Base64.Decoder decoder = Base64.getDecoder();
+
+    /** Constructor which sets the Reader to be decorated.
+     * @param rdr the Reader to be decorated.
+     */
```
Comment (mpsijm) [inline src/main/java/nl/tudelft/ti2206/group9/util/Base64Reader.java]: Please fix the layout of this JavaDoc.

<review, documentation, layout

**Comment by mpsijm** [comment]
Please don't forget to run the SAT checks on your test folder as well. My Checkstyle and PMD were crying. I fixed the warnings for you, but I still have some comments left about your branch. Please resolve them, merge the master and you're good to go :)

<review, logic, documentation

**Comment by mdingjan** [comment]
I reviewed your changes. The util classes Base64Reader and Base64Writer are looking really good!
There aren't any SAT violations left. Please have a look at the comments of Maarten and then this branch is ready to be merged!

<review

File: src/test/java/nl/tudelft/ti2206/group9/util/Base64WriterTest.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    public void tearDown() {
+        try {
+            writer.close();
+        } catch (IOException e) {
+            fail("Base64Writer could not be closed");
+        }
+    }
+
+    @Test
+    public void test() {
```
Comment (mpsijm) [inline src/test/java/nl/tudelft/ti2206/group9/util/Base64WriterTest.java]: Is one test really enough to test the full functionality of the Writer/Reader?

<review, testing

File: src/main/java/nl/tudelft/ti2206/group9/util/Base64Writer.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                    "Base64Writer.writeString()", e.getMessage());
+        } catch (IOException e) {
+            GameObservable.OBSERVABLE.notify(GameObserver.Category.ERROR,
+                    GameObserver.Error.IOEXCEPTION,
+                    "Bas64Writer.writeString(String)", e.getMessage());
+        }
+    }
+
+    @Override
+    public void write(final char[] cbuf, final int off,
```
Comment (mpsijm) [inline src/main/java/nl/tudelft/ti2206/group9/util/Base64Writer.java]: Is this method (and the overriden read method in Reader) tested?

<review, testing

**Comment by mpsijm** [comment]
We'll not be able to fix my comments during this iteration, we'll postpone this to the final iteration.
Please merge.

<review, logic, bug

**Comment by mdingjan** [comment]
No problem, we'll indeed fix your feedback in the last iteration.

Will merge!

<review

**Pull request merged**